### PR TITLE
feat: add `cuckoofilter` to `nippy-jar`

### PR DESCRIPTION
**Pull request by joshieDo**
PR into #4512 

Adds [cuckoofilter](https://github.com/axiomhq/rust-cuckoofilter). There's some docs in the link on why it might be better than bloom filter.

This will be used to try and check what transaction or block hashes are included in the snapshot file (there is a 3% false positive rate), without it being necessarily saved in the data.

### CODE REVIEW

File: crates/storage/nippy-jar/src/lib.rs [hunk 1]
```diff
@@ -308,91 +336,99 @@ impl<'a> NippyJarCursor<'a> {
...
-            vec![3, 13],
-            vec![3, 14],
-            vec![3, 15],
-            vec![3, 16],
-            vec![3, 17],
-            vec![3, 18],
-        ];
+        let (col1, col2) = test_data();
         let num_rows = col1.len() as u64;
         let _num_columns = 2;
```
Comment (gakonst) [inline crates/storage/nippy-jar/src/lib.rs:387]: nit

```suggestion
        let num_columns = 2;
```

File: crates/storage/nippy-jar/src/lib.rs [hunk 2]
```diff
@@ -21,24 +24,24 @@ pub struct NippyJar {
...
     /// soon
-    bloom_filter: bool,
-    /// soon
     phf: bool,
 }
 
 impl NippyJar {
     /// Creates new [`NippyJar`].
-    pub fn new(columns: usize, bloom_filter: bool, phf: bool, path: &Path) -> Self {
+    pub fn new(columns: usize, phf: bool, path: &Path) -> Self {
```
Comment (gakonst) [inline crates/storage/nippy-jar/src/lib.rs:38]: nit would make this `with_phf(bool)` if we expect it to be false most of the time

File: crates/storage/nippy-jar/src/filter/mod.rs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub enum Filters {
+    Cuckoo(Cuckoo),
+    // Avoids irrefutable let errors. Remove this after adding another one.
+    Unused,
+}
+
+impl Filter for Filters {
```
[the remark this answers is not in the document]
Comment (joshieDo) [reply crates/storage/nippy-jar/src/filter/mod.rs:22]: might add bloom later on to benchmark against

File: crates/storage/nippy-jar/src/lib.rs (hunk 2 above)
Comment (joshieDo) [reply crates/storage/nippy-jar/src/lib.rs:38]: overall (parent PR) it's still a draft/wip, it will be removed eventually

**Review by rkrasiuk — APPROVED** [review]
lgtm

File: crates/storage/nippy-jar/src/filter/cuckoo.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+impl PartialEq for Cuckoo {
+    fn eq(&self, _other: &Self) -> bool {
+        self.remaining == _other.remaining &&
+            {
+                #[cfg(not(test))]
+                {
+                    unimplemented!("No way to figure it out without exporting (expensive), so only allow direct comparison on a test")
+                }
```
[the remark this answers is not in the document]
Comment (joshieDo) [reply crates/storage/nippy-jar/src/filter/cuckoo.rs:56]: hm, i feel like that's unsafe behaviour..

File: crates/storage/nippy-jar/src/lib.rs [hunk 5]
```diff
@@ -207,6 +227,14 @@ pub enum NippyJarError {
     ColumnLenMismatch(usize, usize),
     #[error("UnexpectedMissingValue row: {0} col:{1}")]
     UnexpectedMissingValue(u64, u64),
+    #[error("err")]
```
Comment (shekhirin) [inline crates/storage/nippy-jar/src/lib.rs:230]: do we want to do `#[error(transparent)]` here?

**Review by shekhirin — APPROVED** [review]
LGTM, one nit

**Pull request merged**